mesa/agent.py

```diff
@@ -57,3 +60,41 @@ def advance(self) -> None:
     @property
     def random(self) -> Random:
         return self.model.random
+
+
+class AgentSet:
+    def __init__(self, agents: set[Agent], model: Model):
+        self._agents = agents
+        self.model = model
+
+    def __len__(self):
+        return len(self._agents)
+
+    def __iter__(self) -> Iterator[Agent]:
+        return iter(self._agents)
+
+    def __contains__(self, agent: Agent) -> bool:
+        """Check if an agent is in the AgentSet."""
+        return agent in self._agents
+
+    def select(self, filter_func: Callable[[Agent], bool] | None = None) -> AgentSet:
+        if filter_func is None:
+            return AgentSet(set(self._agents), self.model)
+        return AgentSet(
+            {agent for agent in self._agents if filter_func(agent)}, self.model
+        )
+
+    def shuffle(self) -> AgentSet:
+        shuffled_agents = list(self._agents)
+        self.model.random.shuffle(shuffled_agents)
+        return AgentSet(set(shuffled_agents), self.model)
+
+    def sort(self, key: Callable[[Agent], Any], reverse: bool = False) -> AgentSet:
+        sorted_agents = sorted(self._agents, key=key, reverse=reverse)
+        return AgentSet(set(sorted_agents), self.model)
+
+    def do_each(self, method_name: str):
+        for agent in self._agents:
+            getattr(agent, method_name)()
```

Comment:
add support for args and kwargs here, and make it possible to get returns.

So 

def do_each(self, method_name:str, *args, **kwargs): --> list[any]
    return [getattr(agent, method_name)(*args, **kwargs) for agent in self._agents]

Reply:
Agreed, thanks for implementing!